### PR TITLE
Tempo: Add warning message when scope missing in TraceQL

### LINKED ISSUE #75092: TraceQL editor should remind me to add `resource.` or `span.` scope to my queries
**Why is this needed**:

TraceQL queries using scoped attributes (e.g., `resource.region`, `span.payload`) complete much faster than queries where the attributes are not scoped (e.g., `.region`, `.payload`). However, someone who doesn't know much about TraceQL would have no way of knowing that, which means they're likely to write queries with unscoped attributes like `{ .region && status=error }` which return slowly. Time spent waiting on query results is frustrating. 

**What would you like to be added**:

I'd like us to add some sort of indicator to the user who is typing a TraceQL query in the Tempo Explore TraceQL tab that shows up when they put an unscoped attribute in their query. That indicator should give them a message that nudges them to apply a scope to their attribute. 

One idea we'd discussed is highlighting the unscoped attribute by underlining it with a squiggly blue/yello line, and then displaying a message on hover to ask the user to scope it: "Consider adding the `span` or `resource` scope to this attribute to improve query performance.". See screenshot below for an example of sort of unscoped attribute that would trigger the yellow underline being displayed.

Screenshot 2023-09-19 at 7 14 04 AM

**Who is this feature for?**

Any user using Grafana Explore's TraceQL tab to write TraceQL queries, especially a new user who doesn't know much about TraceQL and what scopes are/why they matter.

**Further enhancements**
This could be extended to also show a hint to the user if they use a regex operator e.g. `.my-attribute =~ ".*"` , which are extremely expensive, when a `select(…)` operator would achieve the same result and be significantly faster. More context [here](https://raintank-corp.slack.com/archives/C03Q2B4FSM8/p1698769082656329).

### PR DESCRIPTION
**Pull request by joey-grafana**
**What is this feature?**

Adds a warning message when span/resource scope missing in TraceQL.

**Why do we need this feature?**

Queries are much more performant when a span/resource scope is specified so this lets the user know in a warning message if both span and resource scope is missing from the query.

**Who is this feature for?**

Tempo users.

**Which issue(s) does this PR fix?**:

Fixes https://github.com/grafana/grafana/issues/75092

**Special notes for your reviewer:**

Please check that:
- [ ] It works as expected from a user's perspective.
- [ ] If this is a pre-GA feature, it is behind a feature toggle.
- [ ] The docs are updated, and if this is a [notable improvement](https://grafana.com/docs/writers-toolkit/writing-guide/contribute-release-notes/#how-to-determine-if-content-belongs-in-a-whats-new-document), it's added to our [What's New](https://grafana.com/docs/writers-toolkit/writing-guide/contribute-release-notes/) doc.

### CODE REVIEW

**Review by aocenas — APPROVED** [review]
Nice feature 👍

**Comment by aocenas** [comment]
Maybe a small thing though, I wonder if it would make sense to underline the whole identifier instead of just the leading dot, seems like may be easier to miss it this way.

Screenshot 2024-01-17 at 16 51 29

File: public/app/plugins/datasource/tempo/traceql/errorHighlighting.ts [hunk 1]
```diff
@@ -108,40 +112,89 @@ export const getErrorNodes = (query: string): SyntaxNode[] => {
  * Use red markers (squiggles) to highlight syntax errors in queries.
  *
  */
-export const setErrorMarkers = (
+export const setMarkers = (
```
[the remark this answers is not in the document]
Comment (joey-grafana) [reply public/app/plugins/datasource/tempo/traceql/errorHighlighting.ts:115]: Ahh, good catch!

**Comment by joey-grafana** [comment]
> Maybe a small thing though, I wonder if it would make sense to underline the whole identifier instead of just the leading dot, seems like may be easier to miss it this way.
> 
> Screenshot 2024-01-17 at 16 51 29

Yeah honestly this is something I thought about too. The reason I went with just a single char warning message before the dot instead of the entire identifier is that I didn't want to highlight the warning _too much_. As in, it can actually be ignored, especially in the environment is small or in dev where adding the scope wouldn't make a huge difference in terms of performance. I felt like from a UX perspective, it would be better to reserve full identifier highlighting for errors but that we could adapt this approach if/when we get feedback. Wdyt?

**Review by fabrizio-grafana — APPROVED** [review]

**Pull request merged**